✨ feat : 체험 프로그램 생성 API 구현 및 도메인 구조 변경

j-hann opened the pull request:
🟪 Program, ProgramDay, ProgramScheduleTime 코드 리팩토링
 - `null` 값 이슈 해결을 위해 `ProgramDay`, `ProgramScheduleTime`에서 `programId` 직접 주입 제거
 - `Program` Entity `addProgramDay()`, `addProgramScheduleTime()` 메서드에서 `programId` 주입 제거
 - JPA가 `programId` FK를 자동 주입하도록 개선

🟪 Program, AptitudeCategory, RegionCategory 관계 변경
   - `OneToOne`에서 `ManyToOne`으로 변경
   - 하나의 적성 & 지역 카테고리가 0개 또는 N개의 체험 프로그램을 가질 수 있게 재설정 

🟪 AptitudeCategoryRepository , RegionCategoryRepository 생성

🟪 ProgramDayResponseDto, ProgramScheduleTimeResponseDto 생성

🟪 ProgramDayRepository, ProgramScheduleTimeRepository 생성

🟪 RequestDto `@Builder` 제거 후 `@NoArgsConstructor` 추가

🟪 ProgramController 체험 프로그램 생성 API 구현
   - `@AuthenticationPrincipal` 활용해 `JWT` 토큰에서 사용자 정보 추출 로직 추가

🟪 ProgramCreateResponseDto에 `from()` 정적 메서드 추가
   - `programDays`, `programScheduleTimes` 데이터 포함되도록 추가
 
🟪 ProgramService createProgram 메서드 로직 구현
   - `memberId` 기반 `Member` 조회 및 승인된 로컬 크리에이터 체크
   - 적성 & 지역 카테고리 `id` 기반 조회 및 검증
   - 가격 및 할인율 검증, 할인 가격 계산 로직 추가
   - 최소 & 최대 정원 검증 로직 추가
   - 시작일 & 종료일 날짜 범위 검증 메서드 분리
   - Program Entity 생성 & 상태값 기본 설정
   - `ProgramDay`, `ProgramScheduleTime` Entity 생성 및 매핑 로직 추가
      - `시작 시간` + `소요 시간` 통해서 `종료 시간` 자동 계산
   - `ProgramCreateResponseDto.from()`으로 응답 생성

🟪 LocalCreatorRepository, MemberRepository 메서드 추가
   - 중복되는 `.orElseThrow()` 로직을 방지하기 위해 공통 메서드로 추출
   - 서비스 단에서 메서드 호출로 예외 처리까지 적용

🟪 ErrorCode 추가

🟪 Program.http 체험 프로그램 등록 API 호출 스크립트 작성
   - File 첨부 제외 (추후 추가 예정)
   - 로그인 시 `accessToken`을 전역 변수로 저장하도록 스크립트 코드 추가

🟪 Swagger `@SecurityRequirement(name = "Bearer Authentication")` 설정 추가

🟪 ProgramCreateRequestDto `title`, `description`, `curriculumDescription`, `location` 필드 `@NotNull`로 변경

🟪 ProgramService 불필요한 주석 삭제

🟪 AptitudeCategory Entity 적성 코드 필드 추가

🟪  Program Entity discountPrice 필드 finalPrice로 필드명 변경 & ProgramService createProgram() 로직 수정
   - 할인 없으면 price 값 finalPrice으로 넣어주는 로직으로 변경

j-hann committed:
✨  feat : program, category entity 생성, status enum 생성

- program entity 생성

j-hann committed:
✨  feat : 체험 프로그램 등록 API

- controller 로직 추가, DTO 생성, service 로직 추가, http 파일 추가

j-hann committed:
🔀  branch : user branch 병합

- user branch랑 병합

j-hann committed:
✨  feat : 체험 프로그램 생성 API

-

j-hann committed:
Merge branch 'develop' into feature/program

j-hann committed:
✨  feat : 체험 프로그램 생성 API

j-hann committed:
Merge branch 'develop' into feature/program

j-hann committed:
✨  feat : 체험 프로그램 생성 API 관련 구조 수정 및 구현

- Program entity 수정, ProgramDay & ProgramScheduleTime entity 생성, Dto 수정 및 생성, swagger 설명 추가

j-hann committed:
Merge branch 'develop' into feature/program

j-hann committed:
♻️  refactor : AptitudeCategory Table명 수정, repository 타입 명시

- validation 의존성 추가, ProgramService 체험 프로그램 생성 준비 코드 추가

j-hann committed:
🔧  add : Program entity, originalProgram 필드 nullable 설정 추가

- originalProgram 필드 null값 허용 명시

j-hann committed:
Merge branch 'develop' into feature/program

j-hann committed:
Merge branch 'develop' into feature/program

j-hann committed:
Merge branch 'develop' into feature/program

j-hann committed:
Merge branch 'develop' into feature/program

j-hann committed:
✨  feat : 체험 프로그램 생성 API 구현

- Program entity 모집기간 필드 삭제, - 요일, 프로그램 스케줄 시간 orphanRemoval false로 변경, createProgram 서비스 로직 추가

j-hann committed:
🔀  branch : develop merge

- develop branch merge

j-hann committed:
✨  feat : 체험 프로그램 생성 API 구현 및 도메인 구조 수정

- Program, ProgramDay, ProgramScheduleTime Entity 연관관계 수정

j-hann committed:
♻️  refactor : Program 연관 Entity 구조 변경, Dto 코드 수정 및 추가

- Program, ProgramDay, ProgramScheduleTime 코드 리팩토링, - Program Entity AptitudeCategory, RegionCategory 관계 변경, - ProgramCreateRequestDto title, description, curriculumDescription, location 필드 @NotNull로 변경, - ProgramCreateResponseDto에서 from() 메서드에 programScheduleTimes id 추가, Program.http 로그인 시 accessToken을 전역 변수로 저장하도록 스크립트 코드 추가, - 불필요한 주석 삭제

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/program/service/ProgramService.java:
```diff
@@ -36,11 +52,128 @@ public ProgramService(AptitudeCategoryRepository aptitudeCategoryRepository, Reg
+			throw new CustomException(ErrorCode.INVALID_PRICE_RANGE);
+		}
+
+		BigDecimal discountedPrice = null;
```
현재 `Program` 엔티티에서 `discountedPrice`가 `nullable`로 되어 있어, 할인이 없는 경우 `null`이 들어가는데  
할인된 가격보다 할인이 적용된 후 최종 가격을 별도의 `finalPrice` 필드로 계산해 저장하는 게 더 명확할 것 같아요!!!

할인이 없으면 그냥 `BigDecimal finalPrice = price;`로 하게 되면 `null` 체크 없이도 로직이 단순해지고 값 처리가 편할 것 같습니다.

j-hann replied:
짱짱👍

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/program/controller/ProgramController.java:
```diff
@@ -4,17 +4,18 @@
 import org.springframework.web.bind.annotation.RequestBody;
-import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RestController;
 
```
@Tag 달아주시면 안될까요?🥺

j-hann replied:
안됩니다..

j-hann replied:
사실 됩니다..!👍

j-hann committed:
🔧  add : add : 적성 카테고리 entity, Program 필드 finalPrice로 변경, ProgramService createProgram() 로직 수정

- 적성 카테고리 entity, Program 필드 finalPrice로 변경, ProgramService createProgram() 로직 수정

j-hann committed:
Merge branch 'develop' into feature/program

j-hann committed:
🔧  add : ProgramCreateResposeDto 필드 변경

- discountPrice에서 finalPrice로 필드명 변경
